apps/api/src/app/events/usecases/add-job/add-digest-job.usecase.ts

```diff
@@ -1,55 +1,100 @@
-    protected createExecutionDetails: CreateExecutionDetails
-  ) {}
-
-  public async execute(command: AddJobCommand): Promise<number | undefined> {
-    const data = await this.jobRepository.findById(command.jobId);
-    if (!data) throw new ApiException('Job not found');
-
-    const isValidDigestStep = data.type === StepTypeEnum.DIGEST && data.digest?.amount && data.digest.unit;
-    if (!isValidDigestStep || !data) {
-      return undefined;
+  constructor(private jobRepository: JobRepository, protected createExecutionDetails: CreateExecutionDetails) {}
+
+  public async execute(command: AddDigestJobCommand): Promise<AddDigestJobResult> {
+    const { job } = command;
+
+    this.validateDigest(job);
+
+    return await this.shouldDelayDigestOrMerge(job);
+  }
+
+  private validateDigest(job: JobEntity): void {
+    if (job.type !== StepTypeEnum.DIGEST) {
+      throw new ApiException('Job is not a digest type');
     }
 
-    const shouldAddDigest = await this.shouldAddDigestJob.execute(command);
+    if (!job.digest?.amount) {
+      throw new ApiException('Invalid digest amount');
+    }
+
+    if (!job.digest?.unit) {
+      throw new ApiException('Invalid digest unit');
+    }
+  }
 
-    if (!shouldAddDigest) {
-      await this.createExecutionDetails.execute(
-        CreateExecutionDetailsCommand.create({
-          ...CreateExecutionDetailsCommand.getDetailsFromJob(data),
-          detail: DetailEnum.DIGEST_MERGED,
-          source: ExecutionDetailsSourceEnum.INTERNAL,
-          status: ExecutionDetailsStatusEnum.SUCCESS,
-          isTest: false,
-          isRetry: false,
-        })
-      );
+  private async shouldDelayDigestOrMerge(job: JobEntity): Promise<AddDigestJobResult> {
+    const { matched, modified } = await this.shouldDelayDigestOrMergeWithLock(job);
+
+    // We merged the digest job as there was an existing delayed digest job for this subscriber and template in the same time frame
+    if (matched > 0 && modified === 0) {
+      await this.digestMergedExecutionDetails(job);
 
       return undefined;
     }
 
-    await this.jobRepository.updateStatus(command.organizationId, data._id, JobStatusEnum.DELAYED);
-    if (!data.digest?.amount || !data.digest?.unit) {
-      throw new ApiException('Invalid digest amount or unit');
+    // We delayed the job and created the digest
+    if (matched === 0 && modified === 1) {
+      const { digest } = job;
+
+      if (!digest?.amount || !digest?.unit) {
+        throw new ApiException(`Somehow ${job._id} had wrong digest settings and escaped validation`);
+      }
+
+      return AddJob.toMilliseconds(digest.amount, digest.unit);
     }
 
-    return AddJob.toMilliseconds(data.digest.amount, data.digest.unit);
+    return undefined;
+  }
+
+  private async shouldDelayDigestOrMergeWithLock(job: JobEntity): Promise<IFindAndUpdateResponse> {
+    const resource = `environment:${job._environmentId}:template:${job._templateId}:subscriber:${job._subscriberId}`;
```

Comment:
Opinionated structure. But the key params that identify if we have to merge digests that happen at the same time is based on these 3 values: `environmentId`, `templateId` and `subscriberId`.
If not using them we would expect faulty behaviour of the lock and the digest implementation.

Reply:
Should we also add here the `digestKey` as part of the lock? Since different digest keys might not affect each other and other is no need to lock them.

Reply:
I can temporarily do it in case `digestKey` is provided. But I need to revisit the `digestKey` feature after this anyway. There have been feedback is not working properly.

Reply:
So let's leave it for later than